File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-2.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # handle the rotation by finding deflection point

        l, r = 0, len(nums) - 1
        
        while l < r:
            mid = l + ((r - l) // 2)

            if nums[mid] > nums[r]:
                l = mid + 1
            else:
                r = mid

        pivot = l

        # handle the sort

        def binary_search(left: int, right: int) -> int:
            while left <= right:
                mid = left + ((right - left) // 2)

                if target < nums[mid]:
                    right = mid - 1
                elif target > nums[mid]:
                    left = mid + 1
                else:
                    return mid

            return -1
        
        if target >= nums[pivot] and target <= nums[len(nums) - 1]:
            return binary_search(pivot, len(nums) - 1)
        else:
            return binary_search(0, pivot - 1)    

File: Data Structures & Algorithms/find-peak-element/submission-2.java
class Solution {
    public int findPeakElement(int[] nums) {
        int l = 0;
        int r = nums.length - 1;
        while (l <= r) {
            int peakCandidate = (l + r) / 2;
            if (peakCandidate > 0 && nums[peakCandidate] <= nums[peakCandidate - 1]) {
                r = peakCandidate - 1;
            } else if (peakCandidate < nums.length - 1 && nums[peakCandidate] <= nums[peakCandidate + 1]) {
                l = peakCandidate + 1;
            } else {
                return peakCandidate;
            }
        }
        return -1;
    }
}